controller/hybridgateway/plugin/plugin.go

```diff
@@ -16,78 +17,135 @@ import (
+//     - If an API error (other than NotFound) occurs, returns the error.
+//     - If NotFound, returns the freshly built plugin (selfManaged=false) for creation.
+//     - If found:
+//     a. Preserves the existing hybrid routes annotation value.
+//     b. Uses AnnotationManager to append the current HTTPRoute reference.
+//     c. Returns the updated (rebuilt) plugin (selfManaged=false).
+//  5. Spec reconciliation for existing plugins is not performed yet (see TODO / issue 2687).
+//
+// Self-managed plugin:
+//
+//	A plugin referenced via ExtensionRef. These are returned as-is without rebuild,
+//	annotation modification, or existence checks.
 //
 // Parameters:
-//   - ctx: The context for API calls and cancellation
-//   - logger: Logger for structured logging
-//   - cl: Kubernetes client for API operations
-//   - httpRoute: The HTTPRoute resource from which the KongPlugin is derived
-//   - filter: The specific filter within the HTTPRoute rule
-//   - pRef: The parent reference (Gateway) for the HTTPRoute
+//
+//	ctx        - Context for API calls.
+//	logger     - Structured logger.
+//	cl         - Kubernetes client.
+//	httpRoute  - Source HTTPRoute.
+//	filter     - The HTTPRouteFilter being processed.
+//	pRef       - Parent (Gateway) reference.
 //
 // Returns:
-//   - *configurationv1.KongPlugin: The created or updated KongPlugin resource
-//   - error: Any error that occurred during the process
+//
+//	*configurationv1.KongPlugin - Built or retrieved plugin.
+//	selfManaged                 - True if sourced from ExtensionRef.
+//	error                       - Any error encountered.
 func PluginForFilter(
 	ctx context.Context,
 	logger logr.Logger,
 	cl client.Client,
 	httpRoute *gwtypes.HTTPRoute,
 	filter gwtypes.HTTPRouteFilter,
 	pRef *gwtypes.ParentReference,
-) (*configurationv1.KongPlugin, error) {
+) (kongPlugin *configurationv1.KongPlugin, selfManaged bool, err error) {
 	pluginName := namegen.NewKongPluginName(filter)
 	logger = logger.WithValues("kongplugin", pluginName)
 	log.Debug(logger, "Generating KongPlugin for HTTPRoute filter")
 
+	// In case the filter is an ExtensionRef, retrieve the referenced KongPlugin and return early
+	if filter.Type == gatewayv1.HTTPRouteFilterExtensionRef {
+		log.Debug(logger, "Filter is an ExtensionRef, retrieving referenced KongPlugin")
+		plugin, err := getReferencedKongPlugin(ctx, cl, httpRoute.Namespace, filter)
+		if err != nil {
+			log.Error(logger, err, "Failed to retrieve referenced KongPlugin")
+			return nil, false, fmt.Errorf("failed to retrieve referenced KongPlugin %s: %w", pluginName, err)
+		}
+		log.Debug(logger, "Successfully retrieved referenced KongPlugin")
+		return plugin, true, nil
+	}
+
 	plugin, err := builder.NewKongPlugin().
 		WithName(pluginName).
 		WithNamespace(httpRoute.Namespace).
 		WithLabels(httpRoute, pRef).
 		WithAnnotations(httpRoute, pRef).
-		WithFilter(filter).
+		WithFilter(ctx, cl, httpRoute.Namespace, filter).
```

Comment:
This change can be dropped (we don't really use the extra args anymore).
```suggestion
		WithFilter(filter).
```